src/macros-by-example.adoc

```diff
@@ -0,0 +1,826 @@
+“escaping” it with `*literal*`.
+
+TIP: Our documents use bold typewriter face to distinguish special forms and keywords from symbols
+referencing macros and parameters.
+
+The critical point is that special forms are “special” precisely because they cannot be expressed
+as macros and must therefore receive bespoke syntactic treatment.  Since the elements of
+macro-invocation expressions are themselves expressions, when you want something to _not_ be
+evaluated that way, it must be a special form.
+
+Finally, these special forms are part of the template language itself, and are not visible to
+encoded data: the E-expression `(:literal foo)` must necessarily refer to some user-defined macro
+named `literal`, not to this special form.  (There’s also no need for such a form in
+E-expressions, because in that context symbols and S-expressions are not “evaluated”, and
+everything is literal except for E-expressions (which are not data, but encoding artifacts).)
+
+
+=== Parameter Types
+
+In our examples so far, the macro signatures have been simple lists of parameter names, and each
+parameter accepts a value of any type.  But this is often undesirable, since the resulting output
+could be schematically malformed, or worse(?), macro-expansion could fail in hard-to-diagnose ways:
+
+[{mrk}]
+----
+(:detail_page_url [true]) ⇒ _**error**: make_string expects a string_
+----
+
+This E-expression cannot be expanded because `make_string` requires its arguments to be textual
+values, and `[true]` is not a string or symbol.  But this failure happens within the
+implementation of `detail_page_url`, not the point where the error occurred.  In this example
+those points are only one step removed, but it’s not hard to imagine macros where the stack is
+deep enough to make diagnosis quite difficult.
+
+To detect problems close to their source, macro signatures can declare type constraints on their parameters:
+
+[{mrk}]
+----
+(*public* detail_page_url
+  [(*string* asin)]
+  (website_url (make_string "dp/" asin)))
+----
+
+Here we constrain the `asin` parameter to produce a string so the blame is more clear:
+
+[{mrk}]
+----
+(:detail_page_url [true]) ⇒ _**error**: detail_page_url expects a string_
+----
+
+These types also serve a second, perhaps more impactful, purpose: they can allow the binary
+encoding to be more compact by omitting type information or using fixed-width values.
+
+This example reveals additional syntax for parameter declarations.  So far, a parameter was
+declared by a symbol denoting its name, now we have an S-expression pairing a type and a name.
+In this context the types include all the normal “concrete” Ion types, plus some useful abstract
+supertypes like `*number*`, `*text*`, and `*lob*`, as well as the unconstrained “top type” `*any*`.
+The latter is the default type, so the signature `[foo]` is equivalent to `[(*any* foo)]` meaning
+that the parameter `foo` accepts any value.
+
+
+=== Cardinality: Rest Parameters
```

Comment:
As someone who's not familiar with the literature that "rest parameters" probably comes from, it just makes me sleepy when I see it. Maybe we could wake readers up by quickly saying something like "rest parameters, as in: 'all the rest of them'".

Reply:
I think the term doesn't map well to this language's stream-nature, so I avoided it.

"Variadic" is ambiguous here due to the rules around omitting trailing voidable arguments.  It's the _macro_ that's variadic (ie, it accepts a variable number of arguments), due to the cardinality of one or more arguments.

PS it's not too late to change terminology, so keep the critique coming.